[ABLD-184] Use hermetic python 3.12 for Bazel work

aiuto opened the pull request:
- Use 3.12 as the defaul toolchain
- set up pip dev requirements file
  - The development dependencies are specifically separate from the included dependencies we ship. We'll add that list when the time comes.
- provide a smoke test that python works
  - the test is intentionally brittle w.r.t. python versions. When we try to update to 3.13, the test should fail, and we know it really took

Why do we import invoke?
Because I need to turn various libraries under /tasks into py_binaries we can call from bazel. They all import invoke.

https://datadoghq.atlassian.net/browse/ABLD-184

aiuto committed:
Use hermetic python 3.12

- Use 3.12 as the defaul toolchain
- set up pip dev requirements file
  - The development dependencies are specifically separate from the included
    dependencies we ship. We'll add that list when the time comes.
- provide a smoke test that python works
  - the test is intentionally brittle w.r.t. python versions.
    When we try to update to 3.13, the test should fail, and we know it really took

aiuto committed:
run tests in CI

aiuto committed:
fix windows tags

chouquette commented on MODULE.bazel:
```diff
@@ -84,6 +83,33 @@ register_toolchains(
+
+pip = use_extension("@rules_python//python/extensions:pip.bzl", "pip")
+
+# Devopment requirements. These are the tools needed to build the agent.
```
```suggestion
# Development requirements. These are the tools needed to build the agent.
```

chouquette reviewed (commented):
LGTM but I'll let other people with more python knowledge a chance to review before approving

rdesgroppes commented on .gitlab/bazel/tests.yaml:
```diff
@@ -0,0 +1,29 @@
+
+bazel:tests:windows-amd64:
+  extends: [ .bazel:tests, .bazel:ext:windows ]
+  tags: [ "windows-v2:2025", "specific:true" ]
```
Kindly remove tags here:
```suggestion
```
Instead, I propose to adjust `.bazel:ext:windows` if need be (single point of maintenance).

rdesgroppes reviewed (approved):
Thanks!

aiuto commented on MODULE.bazel:
```diff
@@ -84,6 +83,33 @@ register_toolchains(
+python.defaults(python_version = "3.12")
+python.toolchain(
+    is_default = True,
+    python_version = "3.12",
```
let me try that.

aiuto committed:
use .python-version for test

aiuto committed:
typo

aiuto committed:
Merge branch 'main' into aiuto/py1

aiuto committed:
Merge branch 'main' into aiuto/py1

aiuto committed:
Merge branch 'main' into aiuto/py1

aiuto committed:
call

rdesgroppes committed:
Fix `This GitLab CI configuration is invalid`

This is meant to fix:
> This GitLab CI configuration is invalid: jobs:bazel:tests:linux-amd64 config contains unknown keys: cache, tags

(https://gitlab.ddbuild.io/DataDog/datadog-agent/-/ci/editor?branch_name=aiuto%2Fpy1&tab=2)

rdesgroppes commented:
Took me some time to figure out why the tests are failing on Linux/AMD64 because `bazel` only prints:
```
FAIL: //bazel/tests:does_python_work (Exit 127) (see /data/bzl/bazel/649c250ddb88e3a1390a686c647ce4c2/execroot/_main/bazel-out/k8-fastbuild/testlogs/bazel/tests/does_python_work/test.log)
INFO: Found 1 test target...
Target //bazel/tests:does_python_work up-to-date:
  bazel-bin/bazel/tests/does_python_work
INFO: Elapsed time: 12.121s, Critical Path: 0.28s
INFO: 6 processes: 1 remote cache hit, 4 internal, 1 processwrapper-sandbox.
INFO: Build completed, 1 test FAILED, 6 total actions
//bazel/tests:does_python_work                                           FAILED in 0.0s
  /data/bzl/bazel/649c250ddb88e3a1390a686c647ce4c2/execroot/_main/bazel-out/k8-fastbuild/testlogs/bazel/tests/does_python_work/test.log
```
In Bash, an exit code of 127 specifically indicates "command not found."
So I gave a [try](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/commit/cda9e1dd06d68bcb4e22508f3c88a9bba8336e6f) dumping `/data/bzl/bazel/649c250ddb88e3a1390a686c647ce4c2/execroot/_main/bazel-out/k8-fastbuild/testlogs/bazel/tests/does_python_work/test.log`:
```
-----------------------------------------------------------------------------
/usr/bin/env: 'python3': No such file or directory
```
(https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/1206389004#L52)

**We don't seem to use the hermetic Python toolchain here**.

Luckily, the problem surfaces on Linux/AMD64 likely because the persistent runner has no `python3` in the PATH.

aiuto committed:
Merge branch 'main' into aiuto/py1

aiuto committed:
add env toolchain

aiuto committed:
Merge branch 'main' into aiuto/py1

aiuto committed:
the linter is wrong

aiuto committed:
logs do not upload automatically?

aiuto committed:
wildcard

aiuto committed:
try the bootstrapper

aiuto committed:
typo

aiuto committed:
rev

rdesgroppes committed:
Do not bootstrap with host python
